File: BOJ/14888.cpp
#include <iostream>
using namespace std;

int N;
int a[11], op[4];
int Max = -1e9, Min = 1e9;

void cal(int index, int ans, int add, int sub, int mul, int div) {
    if (index == N) {
        if (ans > Max) 
            Max = ans;
        if (ans < Min) 
            Min = ans;
        return;
    }
    if (add) 
        cal(index + 1, ans + a[index], add - 1, sub, mul, div);
    if (sub) 
        cal(index + 1, ans - a[index], add, sub - 1, mul, div);
    if (mul) 
        cal(index + 1, ans * a[index], add, sub, mul - 1, div);
    if (div) 
        cal(index + 1, ans / a[index], add, sub, mul, div - 1);
}

int main() {
    cin>>N;
    for (int i = 0; i < N; i++)
        cin >> a[i];
    for (int i = 0; i < 4; i++)
        cin >> op[i];
    cal(1, a[0], op[0], op[1], op[2], op[3]);
    cout << Max << '\n' << Min << '\n';
    return 0;
}

File: BOJ/14888.py
import math

n = int(input())
nums = list(map(int, input().split()))
pl, sub, mul, div = map(int, input().split())

Min, Max = 1e9, -1e9

def dfs(result, idx, p, s, m, d):
    global Min, Max
    
    if p<0 or s<0 or m<0 or d<0:
        return
    if idx == n:
        Min = min(Min, result)
        Max = max(Max, result)
        return

    dfs(result+nums[idx], idx+1, p-1, s, m, d)
    dfs(result-nums[idx], idx+1, p, s-1, m, d)
    dfs(result*nums[idx], idx+1, p, s, m-1, d)
    dfs(result//nums[idx] if result>0 else -(abs(result)//nums[idx]), idx+1, p, s, m, d-1)
    
dfs(nums[0], 1, pl, sub, mul, div)
print(Max, Min)
